### PR TITLE
[SHELL32] Copy as path shell extension

### PR DESCRIPTION
**Pull request by whindsaks**
Adds the "Copy as path" menu shell extension.

### Notes 
- I could not find suitable existing clipboard code I could use. This new code should be generic enough for anyone to use if need be.
  - This technically violates the documentation regarding EmptyClipboard and a NULL OpenClipboard owner. However, it does work and agrees with [tONT](https://devblogs.microsoft.com/oldnewthing/20210526-00/?p=105252). We simply don't have a HWND to pass.
- The code can be greatly simplified once `SHCreateShellItemArrayFromDataObject` is implemented. 
- Uses IDropTarget instead of IContextMenu for better menu performance.
- Windows always quotes the path, this only quotes it when it contains spaces.

### CODE REVIEW

File: dll/win32/shell32/res/rgs/copyaspathmenu.rgs
Comment (HBelusca) [inline dll/win32/shell32/res/rgs/copyaspathmenu.rgs:3]: Please use 4-space indentation.

**Comment by HBelusca** [comment]
> I could not find suitable existing clipboard code I could use. This new code should be generic enough for anyone to use if need be. 

What about
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/shell32/CDefaultContextMenu.cpp;hb=24b4026ce8aecb9d409b0ad7a378ad7a034bbc66#l921

File: dll/win32/shell32/res/rgs/copyaspathmenu.rgs
Comment (whindsaks) [reply dll/win32/shell32/res/rgs/copyaspathmenu.rgs:3]: Those rgs files all seem to use tabs.

**Comment by whindsaks** [comment]
> What about

No, that is very different. This code only has a dataobject it can read from, it does not have a dataobject to place on the clipboard. The code in shellord.c also does not fit.

**Review by learn-more — APPROVED** [review]
lgtm

File: dll/win32/shell32/CCopyMoveToMenu.cpp
Comment (JoachimHenze) [inline dll/win32/shell32/CCopyMoveToMenu.cpp:464]: is that call to GetLastError() inevitable for proper compatibility? It's such a poor 80ies-like API.

**Review by JoachimHenze — COMMENTED** [review]
Personally I am no fan of adding this functionality into shell32. Win2k3sp2 doesn't have anything like that as far as I am aware of. And there are much more powerful 3rd-party-applications out there, which do offer that thing in a much more advanced manner, e.g.: https://www.bullzip.com/products/cfn/info.php
Why not let the user choose whether he wants to install something like that, instead of forcing it on them (and at the same time bloating our shell component)?

**Review by JoachimHenze — CHANGES_REQUESTED** [review]
Personally I am no fan of adding this functionality into shell32. Win2k3sp2 doesn't have anything like that as far as I am aware of. And there are much more powerful 3rd-party-applications out there, which do offer that thing in a much more advanced manner, e.g.: https://www.bullzip.com/products/cfn/info.php
Why not let the user choose whether he wants to install something like that, instead of forcing it on them (and at the same time bloating our shell component)?
What about making it modular at least? So that it can be at least uninstalled manually at runtime without touching multiple ros source-files.

-> Not directly into shell32 please. Make it a modular component at least.

File: dll/win32/shell32/res/rgs/copyaspathmenu.rgs
Comment (JoachimHenze) [reply dll/win32/shell32/res/rgs/copyaspathmenu.rgs:3]: Maybe the existing ones are not optimal, old, and maybe partially be inherited from Wine, yes. But yours is an entirely new file. So the proper project rules should be applicable here.

File: dll/win32/shell32/CCopyMoveToMenu.h
Comment (JoachimHenze) [inline dll/win32/shell32/CCopyMoveToMenu.h:113]:
```suggestion
class CCopyAsPathMenu
```
There is EOL-whitespace introduced in line 113. Possibly otherwhere too, use NPP to sledgehammer all of them out!

File: dll/win32/shell32/CCopyMoveToMenu.h
Comment (JoachimHenze) [inline dll/win32/shell32/CCopyMoveToMenu.h:128]: iirc the override keyword is not compatible to older toolchains. It is no hard requirement to change your code, as you are not the first one tainting the ros sources with that keyword. **But I don't want you to uncollapse this comment.** This could never be ported back therefore.

File: dll/win32/shell32/CCopyMoveToMenu.cpp
Comment (whindsaks) [reply dll/win32/shell32/CCopyMoveToMenu.cpp:464]: In theory you could just force return E_FAIL. Who knows if the ultimate caller will care about GlobalAlloc out of memory vs OpenClipboard failure.

File: dll/win32/shell32/lang/tr-TR.rc
Comment (whindsaks) [reply dll/win32/shell32/lang/tr-TR.rc:809]: Windows has & there but I can remove it from all non-English files.

**Comment by whindsaks** [comment]
> Not directly into shell32 please

Windows has it there, added in Vista I think.

File: dll/win32/shell32/CCopyMoveToMenu.h
Comment (whindsaks) [reply dll/win32/shell32/CCopyMoveToMenu.h:128]: I just followed the existing style.

Why did you not say anything when you approved [this PR](https://github.com/reactos/reactos/pull/5873) 5 months ago?!

File: dll/win32/shell32/CCopyMoveToMenu.h
Comment (whindsaks) [reply dll/win32/shell32/CCopyMoveToMenu.h:113]: I do have whitespace visible, just missed that one, thanks.

File: dll/win32/shell32/lang/tr-TR.rc
Comment (binarymaster) [reply dll/win32/shell32/lang/tr-TR.rc]: Do not remove mnemonic keys from non-English resources, this could make translators forget adding them while translating.

File: dll/win32/shell32/CCopyMoveToMenu.h
Comment (JoachimHenze) [reply dll/win32/shell32/CCopyMoveToMenu.h]: Because I give my approval when that PR consisted only of the single commit:
https://github.com/reactos/reactos/pull/5873/commits/526d42168018fc57338043994ab498f78733a78c
Everything that came afterwards had nothing to do with my approval. And my approval has not been reset by Katayama when he pushed more things on top of that. I will and can also not port back all that stuff of Katayama in its current state, for the same reason. Correct.

Like I said: I didn't even request you to change your code here. I just requested you to not collapse this discussion. You are free to act on or ignore this specific review-comment of mine.

File: dll/win32/shell32/lang/tr-TR.rc
Comment (JoachimHenze) [reply dll/win32/shell32/lang/tr-TR.rc]: Why do we even need any accelerators for functions that do not even exist in the target 2k3sp2 even? They do have the power to collide and therefore break the accelerators of features that do actually exist in our target, and they might also break 3rd party stuff that wants to add itself here, and that might come with its own accelerators.
This is adding high chances of future maintenance hell. Avoidable.

File: dll/win32/shell32/CCopyMoveToMenu.cpp
Comment (DarkFire01) [inline dll/win32/shell32/CCopyMoveToMenu.cpp:495]: Can you wrap this in a Vista+ version check instead of just disabling it entirely like this.
The WIN32_VISTA_etc one will be good enough.

**Review by DarkFire01 — APPROVED** [review]
I literally use this feature everyday, super great addition. 
Looks pretty good; I struggle with the variable naming being often these tiny 2 character abbreviations but looking through shell.. that's pretty much how it is often so it makes sense why you did it here.

**Comment by DarkFire01** [comment]
I'll merge this in 48 hours if not one objects, for a reason outside of """Win2k3Sp2"""".
Sorry you had to deal with this, we as a project need to do better.

File: dll/win32/shell32/lang/tr-TR.rc
Comment (DarkFire01) [reply dll/win32/shell32/lang/tr-TR.rc]: This comment is resolved then.

File: dll/win32/shell32/CCopyMoveToMenu.cpp
Comment (whindsaks) [reply dll/win32/shell32/CCopyMoveToMenu.cpp:495]: `SHCreateShellItemArrayFromDataObject` does not actually exist yet in ROS Shell32 so that is why I disabled it outright. I can probably implement IShellItemArray later when other stuff starts using it.

**Comment by whindsaks** [comment]
> consider creating a proper, generalized way for the user to pick-and-choose which items should be shown in the context menus

This is just a plain static verb and can be toggled with tools like https://www.nirsoft.net/utils/shell_menu_view.html

**Comment by DarkFire01** [comment]
> > I'll merge this in 48 hours if no one objects, for a reason outside of """Win2k3Sp2"""". Sorry you had to deal with this, we as a project need to do better.
> 
> @DarkFire01 Not really an _objection_, but...something of note crossed my mind after reading @JoachimHenze 's reservations.
> 
> Consider the sorry state of context menus in File Explorer on Windows. Prior to Windows 11, any old program could shoehorn in its own entries, and there was no obvious way for the user to do anything about it. Now, as of Windows 11, this problem is half-mitigated, half-exacerbated by Microsoft's decision to bisect the context menus, resulting in all the old-style entries being shoved into a secondary "advanced" context menu, while the first right-click exposes a new menu which only displays third-party entries via an entirely new extensibility method.
> 
> With all of that in mind...for ReactOS, I think we should consider creating a proper, generalized way for the user to pick-and-choose which items should be shown in the context menus in File Explorer, both for built-in and third-party entries.
> 
> Admittedly, this proposition is way out of scope for PR at hand, but I felt it was worth voicing in some capacity.
> 
> cc @JoachimHenze If the disabled items weren't even loaded at all (thus alleviating any relevant slowdown in opening the context menus), would that address your concerns adequately?
> 
> cc @jeffbox12 There is design work which would need to be done, should this be pursued. I'll probably need your guidance on how to be..."Reactive™" about that, so to speak.

We should still have sane defaults out of box.
This is a sane default.

but yeah a utility like that would be neat :) good call out.

File: dll/win32/shell32/CCopyMoveToMenu.cpp
Comment (HBelusca) [inline dll/win32/shell32/CCopyMoveToMenu.cpp:471]: What's the idea behind this? For sure, cch * 2 (which is == size) will always be > cch .
Or is it somehow in order to check some sort of value wrapping?

File: dll/win32/shell32/lang/el-GR.rc
Comment (HBelusca) [inline dll/win32/shell32/lang/el-GR.rc:800]: No keyboard accelerator here? (some other translations have it)

**Review by HBelusca — COMMENTED** [review]
Here are some last questions before I approve the PR.

File: dll/win32/shell32/CCopyMoveToMenu.cpp
Comment (learn-more) [reply dll/win32/shell32/CCopyMoveToMenu.cpp:471]: The error message returned in the else already answers this.

**Comment by JoachimHenze** [comment]
@Splitwirez expressed some interesting thoughts here.
Regarding my rejection: I will withdraw that as soon as all accelerators are gone, also in the english files. That is the most defensive approach that we can do here for now.

Actually i still think we shouldn't implement this directly in shell32, but I won't insist on that. After the accelerators will be removed I will turn my rejection into neutrality. I don't want to be an asshole.

**Comment by whindsaks** [comment]
> as soon as all accelerators are gone

Even though I did not invent it, Microsoft did. That makes no sense to me but OK.

File: dll/win32/shell32/lang/el-GR.rc
Comment (whindsaks) [reply dll/win32/shell32/lang/el-GR.rc:800]: The original had them but I removed them because it was unknown if it would clash with other language string accelerators. And now I have been asked to remove the English one as well.

Comment (JoachimHenze) [reply dll/win32/shell32/lang/el-GR.rc:800]: The reason for that request is that we are already in super-clusterfuck-land regarding those accelerators, even for the very easy languages like en-US, en-GB, ko-ko, jp-jp and 3xchinese.
The situation is even worse than that for other languages, see for reference
image

I just don't want to make our life even harder by adding accelerators that do not even exist on 2k3sp2!
I do hope you understand my motivation better now!

Comment (whindsaks) [reply dll/win32/shell32/lang/el-GR.rc:800]: I can understand that. But at the same time, any 3rd-party application can add its own entries here with whatever accelerator it wants. IContextMenu is probably supposed to deal with this somehow.

Comment (JoachimHenze) [reply dll/win32/shell32/lang/el-GR.rc:800]: > any 3rd-party application can add its own entries here with whatever accelerator it wants

Yes, but if a 3rd-party-application decides to not be compatible to 2k3sp2 by not caring or being sloppy, then it is the 3rd-party-apps decision/fault. And not ours.
But we should not punish those apps, that do offer proper accelerators with excellent compat towards our target by doing random stuff. Then we would break the apps, that are actually written correctly and with great care.

**Comment by JoachimHenze** [comment]
Neutral now. If you want to merge it now, merge it.

File: dll/win32/shell32/lang/el-GR.rc
Comment (JoachimHenze) [reply dll/win32/shell32/lang/el-GR.rc:800]: Resolved. If you want to merge the PR now, merge it.

File: sdk/include/reactos/shlguid_undoc.h
Comment (TAN-Gaming) [inline sdk/include/reactos/shlguid_undoc.h:143]:
```suggestion
DEFINE_GUID(CLSID_CopyAsPathMenu,          0xF3D06E7C, 0x1E45, 0x4A26, 0x84, 0x7E, 0xF9, 0xFC, 0xDE, 0xE5, 0x9B, 0xE0);
```

**Comment by learn-more** [comment]
> @Splitwirez expressed some interesting thoughts here. Regarding my rejection: I will withdraw that as soon as all accelerators are gone, also in the english files. That is the most defensive approach that we can do here for now.
> 
> Actually i still think we shouldn't implement this directly in shell32, but I won't insist on that. After the accelerators will be removed I will turn my rejection into neutrality. I don't want to be an asshole.

Why the vendetta against accelerators?

**Pull request merged**
